plugins/usd/module/vtkF3DFaceVaryingPointDispatcher.cxx

```diff
@@ -0,0 +1,134 @@
+
+#include "vtkFloatArray.h"
+#include "vtkInformation.h"
+#include "vtkInformationIntegerKey.h"
+#include "vtkInformationVector.h"
+#include "vtkPointData.h"
+
+vtkStandardNewMacro(vtkF3DFaceVaryingPointDispatcher);
+
+vtkInformationKeyMacro(vtkF3DFaceVaryingPointDispatcher, INTERPOLATION_TYPE, Integer);
+
+//------------------------------------------------------------------------------
+vtkF3DFaceVaryingPointDispatcher::vtkF3DFaceVaryingPointDispatcher() = default;
+
+//------------------------------------------------------------------------------
+vtkF3DFaceVaryingPointDispatcher::~vtkF3DFaceVaryingPointDispatcher() = default;
+
+//------------------------------------------------------------------------------
+int vtkF3DFaceVaryingPointDispatcher::RequestData(vtkInformation* vtkNotUsed(request),
+  vtkInformationVector** inputVector, vtkInformationVector* outputVector)
+{
+  vtkPolyData* input = vtkPolyData::GetData(inputVector[0]->GetInformationObject(0));
+  vtkPolyData* output = vtkPolyData::GetData(outputVector->GetInformationObject(0));
+
+  // early exit if all interpolations are "vertex"
+  vtkPointData* originalPointData = input->GetPointData();
+
+  vtkIdType nbArrays = originalPointData->GetNumberOfArrays();
+
+  bool earlyExit = true;
+
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType != 0) // vertex
+    {
+      earlyExit = false;
+      break;
+    }
+  }
+
+  if (earlyExit)
+  {
+    // nothing to do, just return the input
+    output->ShallowCopy(input);
+    return 1;
+  }
+
+  vtkPoints* originalPoints = input->GetPoints();
+  vtkCellArray* originalFaces = input->GetPolys();
+
+  vtkNew<vtkPoints> newPoints;
+  vtkNew<vtkCellArray> newFaces;
+
+  vtkIdType nbCells = originalFaces->GetNumberOfCells();
+  vtkIdType nbConnectivity = originalFaces->GetNumberOfConnectivityIds();
+
+  // the number of output points is the number of total cells connectivity
+  newPoints->SetNumberOfPoints(nbConnectivity);
+
+  vtkPointData* newPointData = output->GetPointData();
+
+  newPointData->ShallowCopy(originalPointData);
+
+  // Use the interpolation type metadata to know if the array must be replaced
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType == 0) // vertex
+    {
+      auto newArray = vtkSmartPointer<vtkDataArray>::Take(
+        vtkDataArray::CreateDataArray(originalArray->GetDataType()));
+      newArray->SetNumberOfComponents(originalArray->GetNumberOfComponents());
+      newArray->SetNumberOfTuples(nbConnectivity);
```

Comment:
when interpType != 0, the array will not have the right number of tuples ? Am I missing something ?

Reply:
When `interpType` is not 0 (face-varying), the input array can be used directly.  
If `interpType` is 0 (vertex), we need to dispatch data on each duplicated cell vertex